p2p/cuprate-p2p/src/sync_states.rs

```diff
@@ -0,0 +1,429 @@
+    NetworkZone,
+};
+use monero_pruning::{PruningSeed, CRYPTONOTE_MAX_BLOCK_HEIGHT};
+use monero_wire::CoreSyncData;
+
+use crate::{client_pool::disconnect_monitor::PeerDisconnectFut, constants::SHORT_BAN};
+
+/// The highest claimed sync info from our connected peers.
+#[derive(Debug)]
+pub struct NewSyncInfo {
+    /// The peers chain height.
+    chain_height: u64,
+    /// The peers top block's hash.
+    top_hash: [u8; 32],
+    /// The peers cumulative difficulty.
+    cumulative_difficulty: u128,
+}
+
+/// A service that keeps track of our peers blockchains.
+///
+/// This is the service that handles finding out if we need to sync and giving the peers that should
+/// be synced from to the requester.
+pub struct PeerSyncSvc<N: NetworkZone> {
+    /// A map of cumulative difficulties to peers.
+    cumulative_difficulties: BTreeMap<u128, HashSet<InternalPeerID<N::Addr>>>,
+    /// A map of peers to cumulative difficulties.
+    peers: HashMap<InternalPeerID<N::Addr>, (u128, PruningSeed)>,
+    /// A watch channel for *a* top synced peer info.
+    ///
+    /// This is guaranteed to hold the sync info of a peer with the highest cumulative difficulty seen,
+    /// this makes no guarantees about which peer will be chosen in case of a tie.
+    new_height_watcher: watch::Sender<NewSyncInfo>,
+    /// The handle to the peer that has data in `new_height_watcher`.
+    last_peer_in_watcher_handle: Option<ConnectionHandle>,
+    /// A [`FuturesUnordered`] that resolves when a peer disconnects.
+    closed_connections: FuturesUnordered<PeerDisconnectFut<N>>,
+}
+
+impl<N: NetworkZone> PeerSyncSvc<N> {
+    /// Creates a new [`PeerSyncSvc`] with a [`Receiver`](watch::Receiver) that will be updated with
+    /// the highest seen sync data.
+    pub fn new() -> (Self, watch::Receiver<NewSyncInfo>) {
+        let (watch_tx, mut watch_rx) = watch::channel(NewSyncInfo {
+            chain_height: 0,
+            top_hash: [0; 32],
+            cumulative_difficulty: 0,
+        });
+
+        watch_rx.mark_unchanged();
+
+        (
+            Self {
+                cumulative_difficulties: BTreeMap::new(),
+                peers: HashMap::new(),
+                new_height_watcher: watch_tx,
+                last_peer_in_watcher_handle: None,
+                closed_connections: FuturesUnordered::new(),
+            },
+            watch_rx,
+        )
+    }
+
+    /// This function checks if any peers have disconnected, removing them if they have.
+    fn poll_disconnected(&mut self, cx: &mut Context<'_>) {
+        while let Poll::Ready(Some(peer_id)) = self.closed_connections.poll_next_unpin(cx) {
+            tracing::trace!("Peer {peer_id} disconnected, removing from peers sync info service.");
+            let (peer_cum_diff, _) = self.peers.remove(&peer_id).unwrap();
+
+            let cum_dif_peers = self
+                .cumulative_difficulties
+                .get_mut(&peer_cum_diff)
+                .unwrap();
```

Comment:
nit: saving 1 letter with `cum_dif` is not worth it imo

https://github.com/search?q=repo%3Amonero-project%2Fmonero%20cum_dif&type=code

Reply:
ngl this was not intentional